File: src/algorithms/heaps/applications/k-closest-points/sources/k-closest-points.go
// K Closest Points to Origin — use a max-heap of size k (by distance²) to find the k nearest points
package heaps

type pointEntry struct {
	dist  int
	point [2]int
}

func distanceSquaredKCP(point [2]int) int {
	return point[0]*point[0] + point[1]*point[1] // @step:initialize
}

func siftUpKCP(heap []pointEntry, currentIdx int) {
	for currentIdx > 0 {
		parentIdx := (currentIdx - 1) / 2 // @step:sift-up
		if heap[currentIdx].dist > heap[parentIdx].dist {
			// @step:compare
			heap[currentIdx], heap[parentIdx] = heap[parentIdx], heap[currentIdx] // @step:heap-swap
			currentIdx = parentIdx                                                 // @step:sift-up
		} else {
			break // @step:compare
		}
	}
}

func siftDownKCP(heap []pointEntry, heapSize int, parentIdx int) {
	for {
		leftIdx := 2*parentIdx + 1  // @step:sift-down
		rightIdx := 2*parentIdx + 2 // @step:sift-down
		largestIdx := parentIdx     // @step:sift-down
		if leftIdx < heapSize && heap[leftIdx].dist > heap[largestIdx].dist {
			// @step:compare
			largestIdx = leftIdx // @step:sift-down
		}
		if rightIdx < heapSize && heap[rightIdx].dist > heap[largestIdx].dist {
			// @step:compare
			largestIdx = rightIdx // @step:sift-down
		}
		if largestIdx == parentIdx {
			break // @step:sift-down
		}
		heap[parentIdx], heap[largestIdx] = heap[largestIdx], heap[parentIdx] // @step:heap-swap
		parentIdx = largestIdx                                                 // @step:sift-down
	}
}

func kClosestPoints(points [][2]int, kValue int) [][2]int {
	heap := []pointEntry{} // @step:initialize

	for _, point := range points {
		dist := distanceSquaredKCP(point) // @step:heap-insert
		if len(heap) < kValue {
			heap = append(heap, pointEntry{dist, point}) // @step:heap-insert
			siftUpKCP(heap, len(heap)-1)                 // @step:sift-up
		} else if len(heap) > 0 && dist < heap[0].dist {
			// Current point is closer than the farthest in heap — replace root
			heap[0] = pointEntry{dist, point}    // @step:heap-extract
			siftDownKCP(heap, len(heap), 0) // @step:sift-down
		}
	}

	result := make([][2]int, len(heap)) // @step:complete
	for idx, entry := range heap {
		result[idx] = entry.point
	}
	return result // @step:complete
}

File: src/algorithms/heaps/applications/k-closest-points/sources/KClosestPoints.cpp
// K Closest Points to Origin — use a max-heap of size k (by distance²) to find the k nearest points
#include <vector>
#include <utility>

typedef std::pair<int, std::pair<int,int>> HeapEntry;

long long distanceSquared(std::pair<int,int> point) {
    return (long long)point.first * point.first + (long long)point.second * point.second; // @step:initialize
}

void siftUp(std::vector<HeapEntry>& heap, int currentIdx) {
    while (currentIdx > 0) {
        int parentIdx = (currentIdx - 1) / 2; // @step:sift-up
        if (heap[currentIdx].first > heap[parentIdx].first) {
            // @step:compare
            std::swap(heap[currentIdx], heap[parentIdx]); // @step:heap-swap
            currentIdx = parentIdx; // @step:sift-up
        } else {
            break; // @step:compare
        }
    }
}

void siftDown(std::vector<HeapEntry>& heap, int heapSize, int parentIdx) {
    while (true) {
        int leftIdx = 2 * parentIdx + 1; // @step:sift-down
        int rightIdx = 2 * parentIdx + 2; // @step:sift-down
        int largestIdx = parentIdx; // @step:sift-down
        if (leftIdx < heapSize && heap[leftIdx].first > heap[largestIdx].first) {
            // @step:compare
            largestIdx = leftIdx; // @step:sift-down
        }
        if (rightIdx < heapSize && heap[rightIdx].first > heap[largestIdx].first) {
            // @step:compare
            largestIdx = rightIdx; // @step:sift-down
        }
        if (largestIdx == parentIdx) break; // @step:sift-down
        std::swap(heap[parentIdx], heap[largestIdx]); // @step:heap-swap
        parentIdx = largestIdx; // @step:sift-down
    }
}

std::vector<std::pair<int,int>> kClosestPoints(std::vector<std::pair<int,int>>& points, int kValue) {
    std::vector<HeapEntry> heap; // @step:initialize

    for (auto& point : points) {
        long long dist = distanceSquared(point); // @step:heap-insert
        if ((int)heap.size() < kValue) {
            heap.push_back({(int)dist, point}); // @step:heap-insert
            siftUp(heap, (int)heap.size() - 1); // @step:sift-up
        } else if (!heap.empty() && (int)dist < heap[0].first) {
            // Current point is closer than the farthest in heap — replace root
            heap[0] = {(int)dist, point}; // @step:heap-extract
            siftDown(heap, (int)heap.size(), 0); // @step:sift-down
        }
    }

    std::vector<std::pair<int,int>> result; // @step:complete
    for (auto& entry : heap) result.push_back(entry.second);
    return result; // @step:complete
}
